### PR TITLE
bpo-29469: Move constant folding at AST level

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue29469

### CODE REVIEW

**Comment by methane** [comment]
I'll check all comments in https://bugs.python.org/review/29469/
Please don't review for a while.

**Comment by rhettinger** [comment]
Overall, this looks great.  The code is remarkably clean.

**Comment by methane** [comment]
`Python/importlib.h` is conflicted very often.  So I'll fix it right before merge.

**Comment by serhiy-storchaka** [comment]
See my comment on the tracker: https://bugs.python.org/issue29469#msg297596

**Comment by serhiy-storchaka** [comment]
Could you please remove `asdl_ct.py` and simplify `ast_opt.c` by getting rid of useless functions like `astfold_boolop`? The AST types are rarely changed, and I think it would be not hard to update the cleaner code. If sometimes we will need the general code generator we could restore it from your patches.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: `Py_UNREACHABLE()` is now used in `compile.c`.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Move `static int` to a separate line. And similarly in other functions below.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: AFAIK in C99 you can write
```c
[Invert] = PyNumber_Invert,
[Not] = unary_not,
...
```

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: `newval` leaked.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c:192]: `newval` can be NULL.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Shouldn't it return 0 in case of KeyboardInterrupt?

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Missed space after `if`. Missed braces.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Move `{` to a new line.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]:
```c
#undef CALL_INT_SEQ
```

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: s/is/are/ ?

File: Lib/test/test_plistlib.py
Comment (serhiy-storchaka) [inline Lib/test/test_plistlib.py]: Good catch! This test can fail in other implementations. It should be fixed in 3.6 too. Opened #4813 for this.

**Review by serhiy-storchaka — COMMENTED** [review]
It may be worth to document this change in What's New. Even without additional optimizations it can optimize more than the peephole optimizer.

File: Python/ast_opt.c
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Only before return.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM if fix a one simple error.

**Comment by serhiy-storchaka** [comment]
Generated files not up to date

**Comment by serhiy-storchaka** [comment]
Great! Many thanks @methane for your work!

Now I'll try to implement additional optimizations.

File: Python/ast_opt.c
Comment (adrian17) [inline Python/ast_opt.c:83]: Looks like this makes [this](https://github.com/methane/cpython/blob/30edda97b4acd07da7a16c3de7726cd5a23148a9/Python/peephole.c#L348) peephole pass unnecessary too?

Comment (methane) [reply Python/ast_opt.c:83]: Ah, you're right.

Comment (methane) [reply Python/ast_opt.c:83]: Created #4863.  Thanks.

**Pull request merged**
